src/body/tee_body.rs

```diff
@@ -0,0 +1,116 @@
+use bytes::Bytes;
+use http_body::{Body, Frame, SizeHint};
+use std::fmt::{self, Debug, Formatter};
+use std::pin::Pin;
+use std::task::{Context, Poll};
+use tokio::io::AsyncWrite;
+
+use super::BoxBody;
+use crate::error::DynError;
+
+// This struct needs refactored to contain a state enum to ensure that
+// we're able to (flush if necessary) and shutdown the sink when body
+// stops producing frames...
+pub struct TeeBody {
+    body: BoxBody,
+    sink: Pin<Box<dyn AsyncWrite + Send + Sync>>,
+    next: Option<Bytes>,
+}
+
+impl TeeBody {
+    pub fn new<T, U>(body: T, sink: U) -> Self
+    where
+        T: Body<Data = Bytes, Error = DynError> + Send + Sync + 'static,
+        U: AsyncWrite + Send + Sync + 'static,
+    {
+        Self {
+            body: BoxBody::new(body),
+            sink: Box::pin(sink),
+            next: None,
+        }
+    }
+
+    #[inline]
+    pub(crate) fn cap(self) -> BoxBody {
+        // I'm considering bringing back a custom impl of `BoxBody` in the
```

Comment:
This is unfortunate but probably necessary. We're making slightly different assumptions compared to the `http-body-util` crate. We're still maintaining compatibility in that any type from the http-body-util crate that is Unpin can also be used as a `RequestBody` or `ResponseBody` by passing it to our `BoxBody` type rather than the one from the `http-body-util` crate. I would totally use the one from the `http-body-util` crate if there is an `into_inner` method added to the `BoxBody` type. There shouldn't be an issue in doing so because moving the pinned box isn't the same as moving data from outside the pinned pointer. The box itself guarantees a stable memory address and I'd prefer not write a function that can recursively box a request or response body every time it's called.

Reply:
After some consideration, I'm thinking that having the struct wrapper around the `Pin<Box<dyn Body>>` is actually preferred here. Pretty sure it's required for the `TeeBody::cap` fn to work properly.

I'm looking forward to writing some examples that use the BodyExt trait in combination with the concrete types provided by this crate.

I think we're doing a good job keeping the API flexible enough to use with other libraries while promoting best practices and simple solutions to common programing tasks.